File: Practice/Backtracking/10_RegularExpressionMatching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        # Top-Down Memorize
        cache = {}
        
        def dfs(i, j):
            
            if (i,j) in cache:
                return cache[(i,j)]
            
            if i >= len(s) and j >= len(p):
                return True
            if j >= len(p):
                return False
            
            match = i < len(s) and (s[i] == p[j] or p[j] == '.')
            
            if (j+1) < len(p) and (p[j+1] == '*'):
                cache[(i,j)] = ( dfs(i, j+2) or             # not to use *
                                (match and dfs(i+1, j)) )  # use *
                return cache[(i,j)]
            
            if match:
                return dfs(i+1, j+1)
            
            return False
        
        return dfs(0,0)

# Optimize
class Solution:
    def isMatch(self, text, pattern):
        if not pattern:
            return not text

        first_match = bool(text) and pattern[0] in {text[0], '.'}

        if len(pattern) >= 2 and pattern[1] == '*':
            return (self.isMatch(text, pattern[2:]) or
                    first_match and self.isMatch(text[1:], pattern))
        else:
            return first_match and self.isMatch(text[1:], pattern[1:])

File: Practice/Backtracking/10_RegularExpressionMatching.cpp
/*
Input: s = "aa", p = "a*"
Output: true
Explanation: '*' means zero or more of the preceding element, 'a'. Therefore, by repeating 'a' once, it becomes "aa".

Input: s = "mississippi", p = "mis*is*p*."
Output: false
*/
class Solution {
public:
    bool isMatch(string s, string p) {
        ls = s.length(), lp = p.length();
        return dfs(s, p, 0, 0);
    }
    bool dfs(string& s, string& p, int si, int pi){
        if(si >= ls && pi >= lp) return true;
        if(pi >= lp) return false;
        bool match = si < ls && (s[si] == p[pi] || p[pi] == '.');
        if(pi + 1 < lp && p[pi+1] == '*'){
            return dfs(s, p, si, pi+2) || 
                (match && dfs(s, p, si+1, pi));
        }
        return match? dfs(s, p, si+1, pi+1) : false;
    }
private:
    int ls, lp;
};

// Optimized Recursive call
class Solution {
public:
    bool isMatch(string s, string p) {
        if(p.empty()) return s.empty();
        if(p[1] == '*')
            return (isMatch(s,p.substr(2)) || !s.empty() && (s[0] == p[0] || p[0] == '.') && isMatch(s.substr(1), p));
        else
            return !s.empty() && (s[0] == p[0] || p[0] == '.') && isMatch(s.substr(1), p.substr(1));
    }
}


// DP 
class Solution {
public:
    bool isMatch(string s, string p) {
        int m = s.size(), n = p.size();
        vector<vector<bool>> dp(m+1, vector<bool>(n+1, false));
        dp[0][0] = true;
        for(int j=0; j < n; j++){
            if(p[j] == '*' && dp[0][j-1]){
                dp[0][j+1] = true;
            }
        }

        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                if(p[j-1] == '.' || p[j-1] == s[i-1])
                    dp[i][j] = dp[i-1][j-1];
                if(p[j-1] == '*'){
                    if(p[j-2] != s[i-1] && p[j-2] != '.'){
                        dp[i][j] = dp[i][j-2];
                    }else{
                        dp[i][j] = (dp[i][j-1] || dp[i-1][j] || dp[i][j-2]);
                    }
                }
            }
        }
        return dp[m][n];
    }
}